src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp

```diff
@@ -25,45 +25,131 @@ SDPAFusion::SDPAFusion() {
+    auto qk_opt_scaled_opt_mask_added = qk_opt_scaled_mask_added | qk_opt_scaled_pre_mask_opt_reshaped;
+    // Optional reshape after adding mask
+    auto qk_post_mask_shape = any_input();
+    auto qk_post_mask_opt_reshaped = optional<ov::op::v1::Reshape>({qk_opt_scaled_opt_mask_added, qk_post_mask_shape});
+
+    auto softmax = makePattern<ov::op::v8::Softmax>({qk_post_mask_opt_reshaped}, {{"axis", "-1"}});
+    auto softmax_shape = any_input();
+    auto softmax_opt_reshaped = optional<ov::op::v1::Reshape>({softmax, softmax_shape});
+
+    auto qkv_base =
+        makePattern<ov::op::v0::MatMul>({softmax_opt_reshaped, v}, {{"transpose_a", false}, {"transpose_b", false}});
+    auto qkv_shape = any_input();
+    auto qkv_reshaped = wrap_type<ov::op::v1::Reshape>({qkv_base, qkv_shape});
+    auto qkv = qkv_reshaped | qkv_base;
 
     auto valid_qk_shapes = [](const std::shared_ptr<ov::op::v0::MatMul>& qk_matmul) {
         auto q_pshape = qk_matmul->get_input_partial_shape(0);
         auto k_pshape = qk_matmul->get_input_partial_shape(1);
 
-        const size_t q_head_size_idx = 3;
-        const size_t k_head_size_idx = qk_matmul->get_transpose_b() ? 3 : 2;
+        // set size idxes to be counted from the end
+        const int64_t q_head_size_idx = -1;
+        const int64_t k_head_size_idx = qk_matmul->get_transpose_b() ? -1 : -2;
 
-        return q_pshape.size() == 4 && k_pshape.size() == 4 && q_pshape[q_head_size_idx].is_static() &&
-               k_pshape[k_head_size_idx].is_static() &&
+        return q_pshape.size() == k_pshape.size() &&
+               (q_pshape.size() == 2 || q_pshape.size() == 3 || q_pshape.size() == 4) &&
+               q_pshape[q_head_size_idx].is_static() && k_pshape[k_head_size_idx].is_static() &&
                q_pshape[q_head_size_idx].get_length() == k_pshape[k_head_size_idx].get_length();
     };
 
-    ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
         const auto& pattern_map = m.get_pattern_value_map();
         if (transformation_callback(m.get_match_root())) {
             return false;
         }
 
-        auto q_node = pattern_map.at(q);
-        auto k_node = pattern_map.at(k);
-        auto v_node = pattern_map.at(v);
+        auto q_node = pattern_map.at(q_base);
+        auto q_node_ps = q_node.get_partial_shape();
+        if (q_node_ps[-1].is_dynamic() || q_node_ps[-3].is_dynamic())
+            return false;
+
+        auto k_node = pattern_map.at(k_base);
+        auto k_node_ps = k_node.get_partial_shape();
+
+        auto v_node = pattern_map.at(v_base);
+        auto v_node_ps = v_node.get_partial_shape();
+        if (v_node_ps[-1].is_dynamic() || v_node_ps[-3].is_dynamic())
+            return false;
+
+        std::shared_ptr<ov::op::v1::Transpose> k_transpose;
+
+        if (v_node_ps[-1] != k_node_ps[-1]) {
```

Comment:
In general specification SDPA allows different head sizes for K and V inputs, so it should probably be a comparison between Q and K